### ISSUE TITLE
Short NUL-only keys incorrectly detected as duplicates in CBOR and SMILE

### ISSUE THREAD

**Issue by DaveCTurner** [FasterXML/jackson-dataformats-binary#312]
The following object apparently cannot be deserialised in CBOR and SMILE:

```
{"\u0000":"","\u0000\u0000":""}
```

The following are base64-encoded copies of its representation in various formats:

```
JSON: eyJcdTAwMDAiOiIiLCJcdTAwMDBcdTAwMDAiOiIifQ==
YAML: LS0tCiJcMCI6ICIiCiJcMFwwIjogIiIK
SMILE: OikKBfqAACCBAAAg+w==
CBOR: v2EAYGIAAGD/
```

As far as I can tell these all look correct, but on deserialization the field names are reported as duplicates:

```
com.fasterxml.jackson.core.JsonParseException: Duplicate field ''
 at [Source: (byte[])":)
�� � �"; line: -1, column: 11]

	at __randomizedtesting.SeedInfo.seed([29F6747C61080477:49491DCB06A6B558]:0)
	at com.fasterxml.jackson.core.json.JsonReadContext._checkDup(JsonReadContext.java:204)
	at com.fasterxml.jackson.core.json.JsonReadContext.setCurrentName(JsonReadContext.java:198)
	at com.fasterxml.jackson.dataformat.smile.SmileParser._handleFieldName(SmileParser.java:1346)
	at com.fasterxml.jackson.dataformat.smile.SmileParser.nextToken(SmileParser.java:374)
	at com.fasterxml.jackson.core.JsonGenerator._copyCurrentContents(JsonGenerator.java:1935)
	at com.fasterxml.jackson.core.JsonGenerator.copyCurrentStructure(JsonGenerator.java:1914)
	at org.elasticsearch.xcontent.json.JsonXContentGenerator.copyCurrentStructure(JsonXContentGenerator.java:396)
```

Relates https://github.com/elastic/elasticsearch/issues/84146

**Comment by cowtowncoder** [comment]
@DaveCTurner first of all, thank you for reporting this and especially contributing a test to reproduce!

I have a suspicion that the issue is with zero-padding of symbol table used; I vaguely recall something similar with JSON (with escaping, null bytes are legal in JSON names too).
But since Smile and CBOR codecs are bit different (with more straight-forward name decoding due to lack of escaping) it is possible that handling has non working edge case.

But the gist of this is that symbol table lookups are keyed by "quads" (`int` containing up to 4 bytes from name; last quad possibly being incomplete). Special padding is used to avoid conflicts for case of "unused" bits at the end; this only occurs in case of raw `0` bytes.
I forget the exact details of how this was handled in JSON but given time should be able to refresh my memory.

Now: I hope to look into this in near future, but I do have bit of a backlog. I just wanted to add a note saying above. :)

**Comment by cowtowncoder** [comment]
Ok, yes, I can reproduce this. Somehow looks like two-null-byte case gets decoded as single null byte, if I got that right.

**Comment by cowtowncoder** [comment]
Interestingly enough, JSON parser from `jackson-core` does NOT suffer from this. Could be due to escaping required, maybe decoding goes through different route.

**Comment by cowtowncoder** [comment]
Hmmmh. Ok... this is tricky. Code in Smile, CBOR codecs does not keep track of "unused" bytes in quads, unlike JSON codec. But we do need to distinguish between "unused" and `0` bytes since only latter is part of actual decoded value.
Value of `0xFF` (or `-1` as signed) byte works because it is not a legal UTF-8 byte; that's what JSON codec uses.

Two ways to go about it, really:

1. Fill last possibly incomplete quad with -1s when starting (so anything unshifted remains so)
2. Pad at the end

JSON codec uses (2) for what that is worth.

**Comment by cowtowncoder** [comment]
@DaveCTurner ah. I was looking at CBOR end of things. There are indeed a few other code paths to consider... partly for different lengths, partly for boundary conditions. So I'll need to improve the tests I think.

I'll see if I can fix this but just for sake of safety, for 2.14 branch. I have a feeling that there's non-trivial chance for regression for either correctness or performance here.
